src/datamodel/builder/Payload.cpp

```diff
@@ -10,3 +10,100 @@
+Payload::Payload(const std::vector<uint8_t>& value_bytes,
+                 v1::UPayloadFormat format) {
+	if (format < v1::UPayloadFormat::UPAYLOAD_FORMAT_UNSPECIFIED ||
+	    format > v1::UPayloadFormat::UPAYLOAD_FORMAT_SHM) {
+		throw std::out_of_range("Invalid byte vector payload format");
+	}
+	std::string serialized(value_bytes.begin(), value_bytes.end());
+	payload_ = std::make_tuple(serialized, format);
+}
+
+// String constructor
+Payload::Payload(const std::string& value, v1::UPayloadFormat format) {
+	if (format < v1::UPayloadFormat::UPAYLOAD_FORMAT_UNSPECIFIED ||
+	    format > v1::UPayloadFormat::UPAYLOAD_FORMAT_SHM) {
+		throw std::out_of_range("Invalid String payload format");
+	}
+	payload_ = std::make_tuple(value, format);
+}
+
+// Move string constructor
+Payload::Payload(std::string&& value, v1::UPayloadFormat format) {
+	if (format < v1::UPayloadFormat::UPAYLOAD_FORMAT_UNSPECIFIED ||
+	    format > v1::UPayloadFormat::UPAYLOAD_FORMAT_SHM) {
+		throw std::out_of_range("Invalid RValue String payload format");
+	}
+	payload_ = std::make_tuple(std::move(value), format);
+}
+
+// Move Serialized constructor
+Payload::Payload(Serialized&& serialized) {
+	if (std::get<1>(serialized) <
+	        v1::UPayloadFormat::UPAYLOAD_FORMAT_UNSPECIFIED ||
+	    std::get<1>(serialized) > v1::UPayloadFormat::UPAYLOAD_FORMAT_SHM) {
+		throw std::out_of_range("Invalid RValue Serialized payload format");
+	}
+	payload_ = std::move(serialized);
+}
+
+// Move constructor
+Payload::Payload(Payload&& other) noexcept
+    : payload_(std::move(other.payload_)) {}
+
+// Copy constructor
+Payload::Payload(const Payload& other) : payload_(other.payload_) {}
+
+// Move assignment operator
+Payload& Payload::operator=(Payload&& other) noexcept {
+	payload_ = std::move(other.payload_);
+	return *this;
+}
+
+// Copy assignment operator
+Payload& Payload::operator=(const Payload& other) {
+	payload_ = other.payload_;
+	return *this;
+}
+
+Payload::PayloadMoved::PayloadMoved(PayloadMoved&& other) noexcept
+    : std::runtime_error(std::move(other)) {}
+
+// PayloadMoved move assignment operator
+Payload::PayloadMoved& Payload::PayloadMoved::operator=(
+    PayloadMoved&& other) noexcept {
+	std::runtime_error::operator=(std::move(other));
+	return *this;
+}
+
+// PayloadMoved copy constructor
+Payload::PayloadMoved::PayloadMoved(const PayloadMoved& other)
+    : std::runtime_error(other) {}
+
+// PayloadMoved copy assignment operator
+Payload::PayloadMoved& Payload::PayloadMoved::operator=(
+    const PayloadMoved& other) {
+	std::runtime_error::operator=(other);
+	return *this;
+}
+
+// build method
+const Payload::Serialized& Payload::build() const {
+	if (std::get<0>(payload_).empty()) {
+		throw PayloadMoved("Payload is empty or has been already moved");
+	}
+	return payload_;
+}
+
+// move method
+[[nodiscard]] Payload::Serialized Payload::move() && {
+	if (std::get<0>(payload_).empty()) {
+		throw PayloadMoved("Payload has been moved");
```

Comment:
This error message doesn't seem to explain what happened.  The payload is empty.

Reply:
Updated the message as "Payload is empty or has been already moved"